### PR TITLE
chore: refactor coordination

### PR DESCRIPTION
**Pull request by spikecurtis**
Refactors the way clients of the Tailnet API (clients of the API, which include both workspace "agents" and "clients") interact with the API.  Introduces the idea of abstract "controllers" for each of the RPCs in the API, and implements a Coordination controller by refactoring from `workspacesdk`.

chore re: #14729

### CODE REVIEW

**Comment by spikecurtis** [comment]
* **#15346** <a href="https://app.graphite.dev/github/pr/coder/coder/15346?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15345** <a href="https://app.graphite.dev/github/pr/coder/coder/15345?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15344** <a href="https://app.graphite.dev/github/pr/coder/coder/15344?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15343** <a href="https://app.graphite.dev/github/pr/coder/coder/15343?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: tailnet/controllers.go [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+	closed       bool
+	errChan      chan error
+	coordinatee  Coordinatee
+	logger       slog.Logger
+	client       CoordinatorClient
+	respLoopDone chan struct{}
+	sendAcks     bool
+}
+
+func (c *basicCoordination) Close(ctx context.Context) (retErr error) {
```
Comment (ethanndickson) [inline tailnet/controllers.go:154]: This context is only ever used to log a warning if it expires - although this is in the original code, it looks a bit awkward. It's more noticeable (and potentially more annoying) now that it's on the interface `CloserWaiter`. Do we need it?

File: tailnet/controllers.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+	Dial(ctx context.Context, r ResumeTokenController) (ControlProtocolClients, error)
+}
+
+// basicCoordinationController handles the basic coordination operations common to all types of
+// tailnet consumers:
+//
+//  1. sending local node updates to the Coordinator
+//  2. receiving peer node updates and programming them into the Coordinatee (e.g. tailnet.Conn)
+//  3. (optionally) sending ReadyToHandshake acknowledgements for peer updates.
+type basicCoordinationController struct {
```
Comment (ethanndickson) [inline tailnet/controllers.go:104]: I am partial to having interface type assertions on these structs (and the ones in the next PRs), even in cases when it's somewhat obvious what interface they implement - but that might be just me.

**Review by ethanndickson — APPROVED** [review]
Nice, lgtm!

File: tailnet/controllers.go (hunk 2 above)
Comment (spikecurtis) [reply tailnet/controllers.go:104]: I'm not sure what you mean.  Like `var _ CoordinationController = &basicCoordinationController` type stuff?

One problem with that is is that is defeats the compiler check that private structs are actually used somewhere.

File: tailnet/controllers.go (hunk 1 above)
Comment (spikecurtis) [reply tailnet/controllers.go:154]: It's not just logging; it's how long we wait for the remote end to hang up before we forcefully close our end.

Comment (ethanndickson) [reply tailnet/controllers.go:154]: 🤦 my bad

**Review by sreya — APPROVED** [review]
LGTM although I'm admittedly not super well versed in this code

**Pull request merged**